libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/SFHybridViewController.m

```diff
@@ -402,36 +402,50 @@ - (NSString *)sfHybridViewUserAgentString
     return userAgentString;
 }
 
-- (NSURL *)frontDoorUrlWithReturnUrl:(NSString *)returnUrl returnUrlIsEncoded:(BOOL)isEncoded createAbsUrl:(BOOL)createAbsUrl
+- (NSURL *)frontDoorUrlWithReturnUrl:(NSString *)returnUrlString returnUrlIsEncoded:(BOOL)isEncoded createAbsUrl:(BOOL)createAbsUrl
 {
+    // Special case: if returnUrlString itself is a frontdoor.jsp URL, parse its parameters and rebuild.
+    if ([returnUrlString containsString:@"frontdoor.jsp"]) {
+        return [self parseFrontDoorReturnUrlString:returnUrlString encoded:isEncoded];
+    }
+    
     SFOAuthCredentials *creds = [SFUserAccountManager sharedInstance].currentUser.credentials;
-    NSString *instUrl = creds.apiUrl.absoluteString;
-    NSString *fullReturnUrl = returnUrl;
-
+    NSURL *instUrl = creds.apiUrl;
+    NSString *fullReturnUrlString = returnUrlString;
+    
     /*
      * We need to use the absolute URL in some cases and relative URL in some
      * other cases, because of differences between instance URL and community URL.
      */
-    if (createAbsUrl && ![returnUrl hasPrefix:@"http"]) {
-        fullReturnUrl = [NSString stringWithFormat:@"%@/%@", instUrl, returnUrl];
-    }
-    if([returnUrl containsString:@"frontdoor.jsp"]) {
-        NSRange r1 = [returnUrl rangeOfString: isEncoded ? @"retURL%3D" : @"retURL="];
-        NSRange r2 = [returnUrl rangeOfString: isEncoded ? @"%26display" : @"&display"];
-        NSRange range = NSMakeRange(r1.location + r1.length, r2.location - r1.location - r1.length);
-        NSString *newReturnUrl = [returnUrl substringWithRange: range];
-        if(isEncoded) newReturnUrl = [newReturnUrl stringByRemovingPercentEncoding];
-        [SFSDKHybridLogger d:[self class] format:[NSString stringWithFormat:@"%@", newReturnUrl]];
-        return [self frontDoorUrlWithReturnUrl: newReturnUrl returnUrlIsEncoded:TRUE createAbsUrl: FALSE];
-    }
-    NSString *encodedUrl = (isEncoded ? fullReturnUrl : [fullReturnUrl stringByURLEncoding]);
-    NSMutableString *frontDoorUrl = [NSMutableString stringWithString:instUrl];
-    if (![frontDoorUrl hasSuffix:@"/"]) {
-        [frontDoorUrl appendString:@"/"];
+    if (createAbsUrl && ![returnUrlString hasPrefix:@"http"]) {
+        NSURLComponents *retUrlComponents = [NSURLComponents componentsWithURL:instUrl resolvingAgainstBaseURL:NO];
```

Comment:
What happens here if `instUrl` itself contains a path component? For example, `instUrl = https://mycommunity.salesforce.com/community1` and `retUrl = <some URL>`. When we use `NSURLComponents` to break it up, would this still work? That was the reason we replaced it with `[NSString stringWithFormat:...` if I'm not mistaken.

Reply:
> What happens here if `instUrl` itself contains a path component?

`NSURLComponents` breaks a URL into all of its constituent parts, on import.  So in that case, `retUrlComponents.path` will be `/community1`.

> and `retUrl = <some URL>`

In this conditional, `retUrlString` is determined to be a path component (as opposed to an absolute URL).

And then as I mentioned, we just do manipulation on the paths themselves, then regurgitate the `NSURLComponents` object as a string at the end of the process.

Reply:
Sounds good. 👍